### ISSUE TITLE
Improvement of pwndbg.config.Parameter

### ISSUE THREAD

**Issue by cebrusfs** [pwndbg/pwndbg#398]
Since these change may need python2 and python3 tests and may break something, it is better that for some people that familiar with python2 python3 feature to fix these.

1. Support format string by implementing `__format__` to use `self.value.__format__` and remove all workaround.
   We will get error if we put `Parameter` in the format string via `%` or `str.format()`. We have to convert to `str` first for the workaround.
2. Covert every `str` or `bytes` from utf8 to unicode at the setter of `Parameter` and remove all workaround.
  Some config (ex. `nearpc-prefix`) can be put utf8 symbols and the length is used to calculate for padding. Currently, we force decode everything from utf8 to unicode and then compute string length. According to python3 convention, everything is unicode by default. If we do that in the setter, then the conversion can be removed.

**Comment by cebrusfs** [comment]
For 2, this should only affect gdb with python2.
I found out that the `Parameter.value` will be automatically transform back to `utf-8` with `str` type. I think it is done by gdb's python implementation. So, convert at `get_set_string` to set `self.value` will not works. Also, according to [GDB Python docs](https://sourceware.org/gdb/onlinedocs/gdb/Parameters-In-Python.html), we can not prevent gdb to assign `self.value`.

I have a workaround that we just modified all getter to return `unicode` instead of `str` and then make `Parameter` consistence with Python 3 convention. 
However, when to convert it? and how to keep `self.value` up to date?
  a. we can convert each time at `getter`, but call `codes.decode` may cost computing resource when we use any getter of Parameter.
  b. we can convert one time to put the converted value in some member like `self._value` and discard the usage of the `self.value`. All getters return `self._value`. But we need to maintain these two values and make sure nobody use `param.value` to assign a new value since we can not make a setter of `self.value`.

@disconnect3d What do you think?

**Comment by cebrusfs** [comment]
New workaround:
I remap the member name using `__setattr__` and `__getattribute__`. The gdb python implementation will let any value from `set` command write to `gdb.Parameter.value`. However, I remap the member name of Parameter by the following rule
* `Parameter.value` will map to `gdb.Parameter._value` (if it is string type, always keep `unicode`)
* `Parameter.raw_value` will map to `gdb.Parameter.value`

All the getter of `Parameter` return the `Parameter.value`, but `str()` casting will convert `unicode` back to `str`.

**Comment by disconnect3d** [comment]
@cebrusfs Do you have it somewhere working? Would like to check it out as this seems a bit overengineered/too complicated.

I guess there is no way to have both `Parameter.value` and `gdb.Parameter.value` in one instance (without containing one in another) and gdb forces us to use inheritance here.

**Comment by anthraxx** [comment]
I think some parts of this approach will not work at all. For instance, we will not get a sufficient way to pad strings via string-format like we really want it.
We also want to pad with multi-char strings to an exact length, this can't be done with python string format... example: #401 so we need to pad with our own implementation to provide a sufficient method for this.
Furthermore len() will not work properly for escape control chars so we get a proper len of printables that we need for exact matching of length and fill-padding:
```python
>>> len(unicode("\x1b[35mABC"))
>>> len(s)
8
>>> print(s)
ABC
```

**Comment by cebrusfs** [comment]
@anthraxx Yep. But in this issue I just want to fix the `Parameter` type. That is, let `Parameter` can be used without any casting manually. Ex. we can use `'%s' % config.bla` instead of `'%s' % str(config.bla)`

For the format string, I don't have a good idea now. If we don't consider the ANSI symbols, we can use format string to do that
```
>>> fmt = u'{data:{fill}{align}{width}}'
>>> fmt.format(data=u"abc", fill=u"\u2500", align="^", width=20)
u'\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500abc\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500\u2500'
>>> print fmt.format(data=u"abc", fill=u"\u2500", align="^", width=20)
────────abc─────────
```

For the colorized string, I may want to create a class `ColorizedString` then handle all operator and behave like a normal string. It may need to store all coloring meta data and apply the formating first and only add the ANSI character when print.

BTW, the python format string did not support to set `fill` as multiple characters. It is ok for banner now since the unicode is considered as one character. Do we have to support to put the multiple printable character in banner? If so, what is the behavior?

**Comment by cebrusfs** [comment]
@disconnect3d Ok, I created a PR #404 for that and also tested in Python 2.

**Comment by cebrusfs** [comment]
Both issues are fixed by #404 so close this issue.
